jax/_src/state/discharge.py

```diff
@@ -48,6 +53,81 @@ def discharge_state(jaxpr: core.Jaxpr, consts: Sequence[Any]
   new_jaxpr, _ , new_consts = pe.trace_to_jaxpr_dynamic(eval_jaxpr, in_avals)
   return new_jaxpr, new_consts
 
+@dataclasses.dataclass
+class Environment:
+  env: Dict[core.Var, Any]
+
+  def read(self, v: core.Atom) -> Any:
+    if type(v) is core.Literal:
+      return v.val
+    assert isinstance(v, core.Var)
+    return self.env[v]
+
+  def write(self, v: core.Var, val: Any) -> None:
+    self.env[v] = val
+
+class DischargeRule(Protocol):
+
+  def __call__(self, in_avals: Sequence[core.AbstractValue], *args: Any,
+      **params: Any) -> Tuple[Sequence[Optional[Any]], Sequence[Any]]:
+    ...
+
+_discharge_rules: dict[core.Primitive, DischargeRule] = {}
+
+def register_discharge_rule(prim: core.Primitive):
+  def register(f: DischargeRule):
+    _discharge_rules[prim] = f
+  return register
+
+def _has_refs(eqn: core.JaxprEqn):
+  return any(isinstance(v.aval, ShapedArrayRef) for v in eqn.invars)
+
+def _eval_jaxpr_discharge_state(jaxpr: core.Jaxpr, consts: Sequence[Any],
```

Comment:
Note to selves: before this PR, `_eval_jaxpr_dischage_state` had like an if/else chain for our three effectful primitives. But now that `for_loop` as a HOP can itself be effectful, it was a good time to refactor into a more general interpreter with `_discharge_rules`.